flink-streaming-java/src/test/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGeneratorWithOperatorAttributesTest.java

```diff
@@ -0,0 +1,368 @@
+                        new OperatorAttributesConfigurableOperatorFactory<>(
+                                new OperatorAttributesBuilder()
+                                        .setOutputOnlyAfterEndOfStream(true)
+                                        .build()),
+                        BasicTypeInfo.INT_TYPE_INFO,
+                        3,
+                        BasicTypeInfo.INT_TYPE_INFO);
+
+        transform2.addInput(transform, null);
+        transform2.addInput(source2.keyBy(x -> x).getTransformation(), x -> x);
+
+        new DataStream<>(env, transform2).sinkTo(new DiscardingSink<>());
+
+        final StreamGraph streamGraph = env.getStreamGraph(false);
+        Map<String, StreamNode> nodeMap = new HashMap<>();
+        for (StreamNode node : streamGraph.getStreamNodes()) {
+            nodeMap.put(node.getOperatorName(), node);
+        }
+
+        assertThat(nodeMap.get("internalSorter").getInputRequirements()).isEmpty();
+        assertThat(nodeMap.get("noInternalSorter").getInputRequirements().get(0))
+                .isEqualTo(StreamConfig.InputRequirement.PASS_THROUGH);
+        assertThat(nodeMap.get("noInternalSorter").getInputRequirements().get(1))
+                .isEqualTo(StreamConfig.InputRequirement.SORTED);
+    }
+
+    private static class StreamOperatorWithConfigurableOperatorAttributes<IN, OUT>
+            extends StreamMap<IN, OUT> {
+        private final OperatorAttributes attributes;
+
+        public StreamOperatorWithConfigurableOperatorAttributes(
+                MapFunction<IN, OUT> mapper, OperatorAttributes attributes) {
+            super(mapper);
+            this.attributes = attributes;
+        }
+
+        @Override
+        public OperatorAttributes getOperatorAttributes() {
+            return attributes;
+        }
+    }
+
+    private static class TwoInputStreamOperatorWithConfigurableOperatorAttributes<IN1, IN2, OUT>
+            extends CoProcessOperator<IN1, IN2, OUT> {
+        private final OperatorAttributes attributes;
+
+        public TwoInputStreamOperatorWithConfigurableOperatorAttributes(
+                OperatorAttributes attributes) {
+            super(new NoOpCoProcessFunction<>());
+            this.attributes = attributes;
+        }
+
+        @Override
+        public OperatorAttributes getOperatorAttributes() {
+            return attributes;
+        }
+    }
+
+    private static class OperatorAttributesConfigurableOperatorFactory<OUT>
+            implements StreamOperatorFactory<OUT> {
+
+        private final OperatorAttributes operatorAttributes;
+
+        public OperatorAttributesConfigurableOperatorFactory(
+                OperatorAttributes operatorAttributes) {
+            this.operatorAttributes = operatorAttributes;
+        }
+
+        @Override
+        public <T extends StreamOperator<OUT>> T createStreamOperator(
+                StreamOperatorParameters<OUT> parameters) {
+            return null;
```

Comment:
Maybe throw error if this patch is not supposed to be called, instead of returning `null`?